frontend/document/document.lua

```diff
@@ -401,17 +407,37 @@ end
 
 function Document:getFullPageHash(pageno, zoom, rotation, gamma)
     return "renderpg|"..self.file.."|"..self.mod_time.."|"..pageno.."|"
-                    ..zoom.."|"..rotation.."|"..gamma.."|"..self.render_mode..(self.render_color and "|color" or "|bw")
+                    ..zoom.."|"
+                    ..rotation.."|"..gamma.."|"..self.render_mode..(self.render_color and "|color" or "|bw")
+                    ..(self.reflowable_font_size and "|"..self.reflowable_font_size or "")
+end
+
+function Document:getPagePartHash(pageno, zoom, rotation, gamma, rect)
+    return "renderpgpart|"..self.file.."|"..self.mod_time.."|"..pageno.."|"
+                    ..tostring(rect).."|"..zoom.."|"..tostring(rect.scaled_rect).."|"
+                    ..rotation.."|"..gamma.."|"..self.render_mode..(self.render_color and "|color" or "|bw")
                     ..(self.reflowable_font_size and "|"..self.reflowable_font_size or "")
 end
 
 function Document:renderPage(pageno, rect, zoom, rotation, gamma, hinting)
-    local hash_excerpt
-    local hash = self:getFullPageHash(pageno, zoom, rotation, gamma)
-    local tile = DocCache:check(hash, TileCacheItem)
-    if not tile then
-        hash_excerpt = hash.."|"..tostring(rect)
-        tile = DocCache:check(hash_excerpt)
+    -- If rect contains a nested scaled_rect object, our caller handled scaling itself (e.g., drawPagePart)
+    local is_prescaled = rect and rect.scaled_rect ~= nil or false
+
+    local hash, hash_excerpt, tile
+    if is_prescaled then
+        hash = self:getPagePartHash(pageno, zoom, rotation, gamma, rect)
+
+        tile = DocCache:check(hash, TileCacheItem)
+    else
+        hash = self:getFullPageHash(pageno, zoom, rotation, gamma)
+
+        tile = DocCache:check(hash, TileCacheItem)
+
+        -- In the is_prescaled branch above, we're *already* only rendering part of the page
```

Comment:
(These random and uneeded blank lines makes this odd to read :))

Reply:
Yeah, we usually have differing opinions on that subject ;).

FWIW, the intent here was to make the distinction between the two different `getPageHash` calls clearer despite their similar names.